protocol.js

```diff
@@ -36,19 +36,19 @@ export default class JPCWebSocket extends JPCProtocol {
    * @param openPublic {boolean} (optional, default false)
    *   If true, allow other computers from the network to connect.
    *   If false, allow only applications on the local host to connect.
-   *   Currently not supported.
    */
   async listen(secret, port, openPublic) {
     assert(typeof (secret) == "string", "Need secret key");
     assert(typeof (port) == "number", "Need port");
-    let server = new WebSocketNode.Server({ port: port });
+    let host = openPublic ? '0.0.0.0' : '127.0.0.1'; // XXX what about IPv6?
+    this._server = new WebSocketNode.Server({ host: host, port: port });
     return new Promise((resolve, reject) => {
-      server.on("connection", async webSocket => {
+      this._server.on("listening", () => resolve(this._server.address().port));
+      this._server.on("connection", async webSocket => {
         try {
           await this.init(webSocket);
-          resolve();
         } catch (ex) {
-          reject(ex);
+          console.log(ex);
```

Comment:
Why not reject? I think we really should tell the caller. Otherwise this hangs?

(If a log, then it would be `console.error(ex)`)

Reply:
```suggestion
          reject(ex);
```

Reply:
The promise was resolved on listen, so this does nothing at all.

Reply:
Sorry, yes, I missed that bit, `console.error` would certainly have been correct here.